src/Pos/Generator/Block/Payload.hs

```diff
@@ -0,0 +1,186 @@
+                                             TxIn (..), TxInWitness (..), TxOut (..),
+                                             TxOutAux (..), TxSigData (..), mkTx)
+import           Pos.Txp.Logic              (txProcessTransaction)
+import           Pos.Txp.Toil.Class         (MonadUtxo (..), MonadUtxoRead (..))
+import           Pos.Txp.Toil.Types         (Utxo)
+import qualified Pos.Txp.Toil.Utxo          as Utxo
+
+----------------------------------------------------------------------------
+-- Tx payload generation
+----------------------------------------------------------------------------
+
+-- | Generates list of distinct ints in the given range [a,b].
+selectDistinct :: (MonadRandom m, MonadIO m) => Int -> (Int,Int) -> m [Int]
+selectDistinct n0 p@(a, b)
+    | b - a < 0 = error $ "selectDistinct: b < a " <> show p
+    | otherwise = selectDistinct' [] n
+  where
+    n = min (b + 1 - a) n0
+    selectDistinct' cur leftN = do
+        leftItems <- take leftN <$> getRandomRs (a, b)
+        let nubbed = ordNub $ cur ++ leftItems
+        if length nubbed < n
+            then selectDistinct' nubbed (n - length nubbed)
+            else pure nubbed
+
+-- | Separates coin into provided number of coins. All resulting coins
+-- are nonzero.
+splitCoins :: (MonadRandom m, MonadIO m) => Int -> Coin -> m [Coin]
+splitCoins n (fromIntegral . coinToInteger -> c)
+    | c < n = error $ "splitCoins: can't split " <> pretty c <>
+                      " on " <> show n <> " parts"
+    | otherwise = do
+          splitPoints <- sort <$> selectDistinct n (0, c - 1)
+          -- calculate length of intervals
+          let amounts = map (\(a,b) -> b - a + 1) $ (0 : splitPoints) `zip` splitPoints
+          pure $ map (unsafeIntegerToCoin . fromIntegral) amounts
+
+-- | State datatype for transaction payload generation
+data GenTxData = GenTxData
+    { _gtdUtxo     :: Utxo
+      -- ^ Utxo as it is.
+    , _gtdUtxoKeys :: V.Vector TxIn
+      -- ^ Keys of 'gtdUtxo' to support random selection by index.
+    }
+
+makeLenses ''GenTxData
+
+instance (Monad m) => MonadUtxoRead (StateT GenTxData m) where
+    utxoGet txIn = uses gtdUtxo $ M.lookup txIn
+
+instance (Monad m) => MonadUtxo (StateT GenTxData m) where
+    utxoPut txIn txOutAux = gtdUtxo . at txIn ?= txOutAux
+    utxoDel txIn = gtdUtxo . at txIn .= Nothing
+
+-- TODO: move to txp, think how to unite it with 'Pos.Arbitrary.Txp'.
+-- | Generate valid 'TxPayload' using current global state.
+genTxPayload ::
+       forall g m. (RandomGen g, MonadBlockGenBase m)
+    => BlockGenRandMode g m ()
+genTxPayload = do
+    utxo <- lift DB.getAllPotentiallyHugeUtxo
+    let gtd = GenTxData utxo (V.fromList $ M.keys utxo)
+    flip evalStateT gtd $ do
+        (a,d) <- lift $ view tgpTxCountRange
+        txsN <- fromIntegral <$> getRandomR (a, a + d)
+        void $ replicateM txsN genTransaction
+  where
+    genTransaction :: StateT GenTxData (BlockGenRandMode g m) ()
+    genTransaction = do
```

Comment:
In this function alone I counted 8 `error` invocations and at least 5 unsafe functions (`V.unsafeIndex`, 3 times `NE.fromList`, and `unsafeIntegerToCoin`).

The way I'm used to read these unsafe things is that they are not meant to go wrong. I.e. if you call `error`, it's a sanity check (or tricking the type system) and it's never meant to happen at run-time. If you use `unsafeIndex`, you expect it to give you the value because you know that the index is valid.

However, I see nearly no comments as to why these `error` calls don't happen or why unsafe functions are safe in this context. It's hard to reason about code written this way.

To be more clear, I just want a short comment before each `error` or `unsafe-*` function, and in each comment a list of the assumptions which make it safe.